### DISCUSSION TITLE
Negative density of species at steady state condition

### DISCUSSION THREAD

**Discussion by ananthanarasimhanj** [shannon-lab/zapdos#201] [Q&A Plasma Chemistry / CRANE] [answered]
I find negative density for some species (shown in below snapshot); this occurs only when the species evolution has attained steady state.  I have already used "scaling" options. Should I also implement the initial density in "log-molar" form? I have been trying to find out procedure here in Zapdos, to find out things to be taken care while implementing the "log-molar" form. I still have following questions towards it:

1) I am assuming that "log-molar form" is simply the natural logarithm of the species density (originally in cm^-3)? or is it the logarithm of ratio of number density of species in a mole and Avogadro constant? 

2) whether the rate constant values should be changed from cm^x/s to log-molar form? 

I have also attached the input and Cross-section data folder that resulted in negative densities. 

[k.zip](https://github.com/shannon-lab/zapdos/files/11884931/k.zip)

[HNN11RH100E0.8.zip](https://github.com/shannon-lab/zapdos/files/11884936/HNN11RH100E0.8.zip)

![image](https://github.com/shannon-lab/zapdos/assets/113285250/4b87efe4-3028-4acd-a765-ffedb6ea3f24)

**Comment by csdechant** [comment, marked as answer]
Hi, I believe I need some clarification before I answer the questions about Zapdos "log-molar" form.

First, I ran the code and it resulted with a high residual for the default convergence tolerance (I see a "convergence" residual of ~1.3e8). My first recommendation is to tighten the nonlinear relative tolerance (e.g `nl_rel_tol = 1e-14` in the `Executioner` Block). 

Second, based on your input file, it doesn't seem you are using any of Zapdos plasma physics and the simulation is purely CRANE reaction physic (which is fine is you don't need to model fluid physics). While the input file didn't have Zapdos Kernels, CRANE can also run logarithmic form of the densities. To change this for the current input file, your need to change `ODETimeDerivative` to `ODETimeDerivativeLog`, add `use_log = true` in the `ScalarNetwork` Block, and change any initial conditions to $N_{IC} = ln(n_{IC})$ (an example of this can be found in `crane/problems/example3_log.i`).

Now, for the question of units. When you are coupling Zapdos and CRANE, you need to use purely standard SI units (so $m^{-3}$ instead of $cm^{-3}$) this is because all universal constants that Zapdos calls (like permittivity of free space, elementary charge, etc.) are in standard SI units. For you questions:

1. "Log-molar" form is the log of the species density normalized to Avogadro's number (e.g. $N_{j}  = ln(n_{j}/N_{A})$ where $n_{j}$ the species density in units $m^{-3}$ and $N_{j}$ is the logarithmic-molar form.
2. No, the reaction rates only need to be standard SI units.

**Reply by ananthanarasimhanj** [reply]
Thanks for your support with detailed explanations. I will try them and update here.

**Comment by ananthanarasimhanj** [comment]
I have incorporated all the changes you have suggested, and I get negative values (in terms of natural log of density of species) when a steady state is approached, which in real density value is approaching Zero which only needs addressing the reaction mechanism. It looks like the problem is addressed. The zip containing the input file and the database folder is attached here. If possible, please let me know if this looks fine in-terms of residuals and other numeric-related, and also the executioner block is modified.

[HNN13RH100E0.8Log.zip](https://github.com/shannon-lab/zapdos/files/11899704/HNN13RH100E0.8Log.zip)

**Reply by ananthanarasimhanj** [reply]
Is there a way to use scaling for initial condition, while using log molar form?

**Reply by cticenhour** [reply]
In the future, please create a new discussion post as this question alone might be useful to others. You can scale within each variable that gets an initial condition through the variable itself (the `scaling` parameter).  I am unaware of any simple scaling parameters in the default set of initial conditions that can be applied in the `ICs` block, but I imagine you can manually scale those as you define them to align with the scaling done in the problem (at least in a constant scaling sense). If you didn't want to manually scale the problem, you might try turning on automatic scaling (`automatic_scaling = true` in your `Executioners` block).

**Reply by ananthanarasimhanj** [reply]
Thanks! I have some follow up question. Should I create a new discussion post?

**Reply by cticenhour** [reply]
Please do!
